### PR TITLE
Moving engineFunctions to Backend.  Closes #482

### LINKED ISSUE #482: Move engineFunctions to Backend

### PR DESCRIPTION
**Pull request by mcovarr**

### CODE REVIEW

File: src/main/scala/cromwell/engine/backend/local/LocalBackend.scala [hunk 1]
```diff
@@ -213,5 +211,21 @@ case class LocalBackend(actorSystem: ActorSystem) extends Backend with SharedFil
...
+    path
+  }
+
   override def executionInfoKeys: List[String] = List(InfoKeys.Pid)
+
+  override def callEngineFunctions(descriptor: BackendCallJobDescriptor): CallEngineFunctions = {
+    val workflowDescriptor = descriptor.workflowDescriptor
+    val callRoot = callRootPath(descriptor)
+    val stdout = callRoot.resolve("stdout")
+    val stderr = callRoot.resolve("stderr")
```
Comment (scottfrazer) [inline src/main/scala/cromwell/engine/backend/local/LocalBackend.scala]: Seems like the above lines are the same for SGE and Local

Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/local/LocalBackend.scala]: and if I do a little work on `SgeCallEngineFunctions`, perhaps the line below too

**Comment by mcovarr** [comment]
@scottfrazer and @Horneth

File: src/main/scala/cromwell/engine/backend/jes/JesBackend.scala [hunk 2]
```diff
@@ -862,4 +872,17 @@ case class JesBackend(actorSystem: ActorSystem)
...
 
   override def executionInfoKeys: List[String] = List(JesBackend.InfoKeys.JesRunId, JesBackend.InfoKeys.JesStatus)
+
+  override def callEngineFunctions(descriptor: BackendCallJobDescriptor): CallEngineFunctions = {
+    val workflowDescriptor = descriptor.workflowDescriptor
+    val key = descriptor.key
+
+    lazy val callGcsPath = key.callRootPathWithBaseRoot(workflowDescriptor, rootPath(workflowDescriptor.workflowOptions))
+
+    lazy val jesStdoutGcsPath = s"$callGcsPath/${jesLogStdoutFilename(key)}"
```
Comment (Horneth) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: `callGcsPath.resolve(jesLogStdoutFilename(key))`

**Comment by mcovarr** [comment]
thumbses?

**Comment by scottfrazer** [comment]
:+1:

File: src/main/scala/cromwell/engine/backend/Backend.scala [hunk 3]
```diff
@@ -68,6 +70,7 @@ object AttemptedLookupResult {
  * Trait to be implemented by concrete backends.
  */
 trait Backend {
+
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/Backend.scala]: :facepunch:

Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/Backend.scala]: Phew.  I'm not sure what the rules on whitespace are here, I was worried I might be on the receiving end of an entirely different hand gesture.

Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/Backend.scala]: http://gifstumblr.com/images/theres-no-emoticon-for-what-im-feeling_1401.gif

**Comment by Horneth** [comment]
:+1:

**Pull request merged**